File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.py
import sys
input=sys.stdin.readline

# 청소 안되었다면 청소
# 주변칸 중 다 청소가 되어있다면 한칸 후진 / 후진 불가면 멈추기
# 주변칸 중 청소가 안되어있는 칸 있으면 반시계방향 회전 및 앞칸이 청소 안되었다면 전진

def clean(r,c,N,M):
  for i in range(4):
    newX=r+xbox[i]
    newY=c+ybox[i]
    if 0<=newX<N and 0<=newY<M and l[newX][newY]==0:
      return False
  return True


def back(r,c,d,N,M):
  if d==0:
    if r+1<N and l[r+1][c]==0 or l[r+1][c]==-1:
      return [r+1,c]
  elif d==1:
    if c-1>=0 and l[r][c-1]==0 or l[r][c-1]==-1:
      return [r,c-1]
  elif d==2:
    if r-1>=0 and l[r-1][c]==0 or l[r-1][c]==-1:
      return [r-1,c]
  else:
    if c+1<M and l[r][c+1]==0 or l[r][c+1]==-1:
      return [r,c+1]
  return [-1,-1]

def find(r,c,d,N,M):
  if l[r][c]==0:
    l[r][c]=-1
    answer[0]+=1
  if clean(r,c,N,M):
    value=back(r,c,d,N,M)
    if value[0]!=-1:
      find(value[0],value[1],d,N,M)
  else:
    for i in range(4):
      d=(d-1+4)%4
      if d==0:
        if r-1>=0 and l[r-1][c]==0:
          return find(r-1,c,d,N,M)
      elif d==1:
        if c+1<M and l[r][c+1]==0:
          return find(r,c+1,d,N,M)
      elif d==2:
        if r+1<N and l[r+1][c]==0:
          return find(r+1,c,d,N,M)
      else:
        if c-1>=0 and l[r][c-1]==0:
          return find(r,c-1,d,N,M)


N,M=map(int,input().split())
r,c,d=map(int,input().split())
l=[]
for i in range(N):
  l.append(list(map(int,input().split())))

xbox=[-1,1,0,0]
ybox=[0,0,-1,1]
answer=[0]
find(r,c,d,N,M)
print(answer[0])

File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.js
const readline=require('readline');
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const [N,M]=input[iter++].split(' ').map(Number)
  const [r,c,d]=input[iter++].split(' ').map(Number)

  const l=[]
  for(let i=0;i<N;i++){
    l.push(input[iter++].split(' ').map(Number))
  }

  const mx=[-1,1,0,0]
  const my=[0,0,-1,1]

  let nowx=r
  let nowy=c
  let dir=d

  let answer=0
  while (true){
    if (l[nowx][nowy]===0){
      answer+=1
      l[nowx][nowy]=2
    }

    let cleanflag=false
    for(let i=0;i<4;i++){
      const newx=nowx+mx[i]
      const newy=nowy+my[i]
      if (0<=newx && newx<N && 0<=newy && newy<M){
        if (l[newx][newy]===0) {
          cleanflag=true
          break
        }
      }
    }

    if (cleanflag){
      dir=(dir+4-1)%4
      if (dir===0) {
        if (l[nowx-1][nowy]===0){
          l[nowx-1][nowy]=2
          answer+=1
          nowx-=1
        }
      } else if (dir===1){
        if (l[nowx][nowy+1]===0){
          l[nowx][nowy+1]=2
          answer+=1
          nowy+=1
        }
      } else if (dir===2){
        if (l[nowx+1][nowy]===0){
          l[nowx+1][nowy]=2
          answer+=1
          nowx+=1
        }
      } else {
        if (l[nowx][nowy-1]===0){
          l[nowx][nowy-1]=2
          answer+=1
          nowy-=1
        }
      }
    } else {
      if (dir===0) {
        if (l[nowx+1][nowy]!==1){
          if (l[nowx+1][nowy]!==2) answer+=1
          l[nowx+1][nowy]=2
          nowx+=1
        } else break
      } else if (dir===1){
        if (l[nowx][nowy-1]!==1){
          if (l[nowx][nowy-1]!==2) answer+=1
          l[nowx][nowy-1]=2
          nowy-=1
        } else break
      } else if (dir===2){
        if (l[nowx-1][nowy]!==1){
          if (l[nowx-1][nowy]!==2) answer+=1
          l[nowx-1][nowy]=2
          nowx-=1
        } else break
      } else {
        if (l[nowx][nowy+1]!==1){
          if (l[nowx][nowy+1]!==2) answer+=1
          l[nowx][nowy+1]=2
          nowy+=1
        } else break
      }
    }
  }

  console.log(answer)
  process.exit()
})

// 0이면 빈칸
// 1이면 벽
